Convert `functional_architecture.rst` to `functional_architecture.md` (#4535)
Scille/parsec-cloud: docs/development/functional_architecture.md

# Parsec client functional components

Note: This is a work in progress.

- [Parsec client functional components](#parsec-client-functional-components)
  - [Mountpoint](#mountpoint)
  - [Gui](#gui)
  - [Core](#core)
  - [Metadata server](#metadata-server)

## Mountpoint

> The `mountpoint` component provide a virtual filesystem that allow other applications to access its contents.
> This is done using the os filesystem's API that follow native request to the component.

Reviewer: ```suggestion
The `mountpoint` component provides a virtual filesystem allowing external applications to access data managed by Parsec. 
Any operation performed on this virtual file system, such as reading a directory or creating a file, goes from the OS filesystem's API through the `mountpoint` component.
```

Author: done

## Gui

The `GUI` component provide to the user a graphical interface and interact directly with the `core` or the os filesystem (but never with the `mountpoint`).

> On contrary to the `mountpoint` component, the `gui` subscribe to events provided by the `metadata server` to notify the user in case of concurrent operation, deletion of a group, etc.

Author: But I see that I don't used `GUI` everywhere :/

> On contrary to the `mountpoint` component, the `gui` subscribe to events provided by the `metadata server` to notify the user in case of concurrent operation, deletion of a group, etc.

Reviewer: I agree when used as acronym. I proposed lowercase because it seemed to refer to the component name (because of the backticks)

## Core

> On top of the 2 previous componenet, the `core` component is responsible for the client side logic and his composed of 5 sub-modules (see figure below):

Reviewer: ```suggestion
The `core` component is responsible for the client side logic and his composed of 5 sub-modules (see figure below):  
```

Author: done

![Core submodules](../figures/parsec_core_submodules.png)

> - The `Virtual File System` (VFS) (`Système de Fichiers Virtuels`): Receive request related to the filesystem originating from the `mountpoint`, thus expose a filesystem like api.
>   To optimize its performance, that component don't push the modifications to the `metadata server`, it only store the encrypted modifications locally on the machine.

Reviewer: Why keeping the French translation?
```suggestion
- The `Virtual File System` (VFS): Receive filesystem-related requests originating from the `mountpoint`, exposing a filesystem like API.
  To optimize its performance, this component doesn't push the changes to the `metadata server`, instead it only stores the encrypted changes locally on the machine.
```

Author: The used schema use french word, that the way I find to do a mapping

Author: done

> - The `Synchronizer` (`Synchroniseur`): That component periodically transfer the locally stored modifications to the `metadata server`.
>   He listen the server for notification about remote modifications and handle version conflict between the local & remote metadata.

Reviewer: ```suggestion
- The `Synchronizer`: This component periodically transfers the locally stored changes to the `metadata server`.
It listens the server for notifications about remote changes and handles version conflicts between local and remote metadata.
```

Author: done

> - The `Identity manager` (`Gestionnaire d’Identités`): Save the user identity in memory (as a session).

Author: > Save the user identity in memory

Do we mean the volatile memory (RAM) or persistent memory (filesystem) ?

Reviewer: Not sure :/

> - The `Identity manager` (`Gestionnaire d’Identités`): Save the user identity in memory (as a session).

Reviewer: ```suggestion
- The `Identity manager`: Save the user's identity in memory (as a session).
```

Author: done

>   The pass-phrase of the user (which isn't stored) concatened with a SALT encrypt the private key of the terminal (`DEVICE_SIG_S_KEY`) and the user private key (`USER_ENC_S_KEY`) (that one is shared between the other terminals used by the user).

Reviewer: ```suggestion
  The user's pass-phrase (which isn't stored) concatenated with a SALT encrypts the device private key (`DEVICE_SIG_S_KEY`) and the user private key (`USER_ENC_S_KEY`) which is shared between the other devices used by the user.
```

Author: done

>   The `DEVICE_SIG_S_KEY` is used to sign a modification. The `USER_ENC_S_KEY` is used to encrypt user's personal metadata.

Author: Would it be simpler to only say

```suggestion
  The `DEVICE_SIG_S_KEY` is used to sign a modification. The `USER_ENC_S_KEY` is used to encrypt the user's manifest
```

?

Reviewer: :+1:

> - `Messaging` (`Messagerie`): Listen for notifications sent by the `metadata server` (the `core` is link to the `metadata server` with a TCP stream).
>   Those messages can:
> 
>   - Ask the `core` to perform a specific action (like when file is shared)
>   - Are purely informative, those are displayed as a notification on the client application.

Reviewer: ```suggestion
- `Messaging`: Listens for notifications sent by the `metadata server` (the `core` is linked to the `metadata server` with a TCP stream).
  These messages can:
  - Ask the `core` to perform a specific action (like when file is shared)
  - Be purely informative, and are thus displayed as a notification on the client application.
```

> - `Sharing` (`Partage`): Handle the sharing operations. When he receive a workspace share event, it decrypt the message and add the entries to the user manifest.
>   If messages are sent but the user ins't logged in, they are keep on the `metadata server` until the user is connected.

Reviewer: ```suggestion
- `Sharing`: Handles the sharing operations. When it receives a workspace share event, it decrypts the message and adds the entries to the user manifest.
  If messages are sent but the user isn't logged in, they are kept on the `metadata server` until the user is connected.
```

Author: Done

## Metadata server

The `metadata server` is a remote server that contains 4 sub-modules:

> - `Messaging` (`Messagerie`): Send techinical notifications to users.
> 
> - `Data/Metadata Storage` (`Stockage des Données/Métadonnées`): Store file's datas (`BLOCKS`) to the configured object storage provider (AWS, OpenStack Swift or others), The metadatas (Versionned Blobs => `VLOBS`) are stored on PostgreSQL.
> 
> - `Notification` (`Notification des Terminaux`): Notify connected clients when a data is modified or when receiving a new message.
> 
> - `Public Key Infrastructure` (PKI) (`Gestionnaire de Clés Publiques`): module that link a user/device identity with its public keys (`USER_ENC_P_KEY`, `DEVICE_ENC_P_KEY`).

Reviewer: ```suggestion
- `Messaging`: Send technical notifications to users.

- `Data/Metadata Storage`: Stores file's data blocks (`BLOCKS`) to the configured object storage provider (AWS, OpenStack Swift or others), The metadata (Versioned Blobs => `VLOBS`) are stored on the server's PostgreSQL database.

- `Notification: Notifies connected clients when a data is modified or when receiving a new message.

- `Public Key Infrastructure` (PKI): Links a user/device identity with its public keys (`USER_ENC_P_KEY`, `DEVICE_ENC_P_KEY`).
```

Author: done
